Bug-Fix issue: 182 Picker in AddMaintainanceView is not working

rramirez-dev opened the pull request:
Updated the vehicle picker to user the vehicle id and not the vehicle object

# What it Does
* Closes #182 
* Describe what your change does
   The fix allows the user to select a vehicle from the 
   vehicle picker in the AddMaintainanceView form that
   is not the first vehicle in the list.

# How I Tested
* Add a list of steps to show the functionality of your feature
For example:
* Run the application
* Tap the (+) button
* See this screen
etc...

1. Launch the App
2. Load the dashboard by navigating to a different tab     and then select the dashboard tab again.
3. Tap the plus sign to add a new event
4. Enter a title
5. Select a vehicle that is not the first one on the list
6. Tap Add

# Notes
* Anything else that should be noted about how you implemented this feature?
* I filtered the vehicles array based on the id of the vehicle that the user selected. I then get the first item in the array and pass it to the MaintainanceEvent view.

# Screenshot
![Simulator Screen Recording - iPhone 15 - 2023-10-20 at 21 49 41](https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/22773854/8014167a-decc-4823-b52a-f9daa923885f)

rramirez-dev committed:
Bug-Fix issue: 182 Picker in AddMaintainanceView is not working

Updated the vehicle picker to user the vehicle id and not the vehicle object

mikaelacaron reviewed (commented):
Can you make this work using the `Vehicle` object as opposed to the id? if not use the id and change the variable to `selectedVehicleID`
Also there's some indentation that's incorrect, your Xcode should be set to `4` spaces

rramirez-dev committed:
PR-182-Review: Enable picker in AddMaintainanceView to use a vehicle object

-As per the documentation in order to allow the picker to user an object the tag object
 has to match the selectedVehicle type

mikaelacaron reviewed (changes requested):
nevermind sorry for the confusion, go ahead and use the `id`, that's how it's done in the `AddOdometerReadingView`

Also note you don't need the `vehicle as Vehicle` anymore because of that, and technically with the current way that you fixed this the `if !vehicles.isEmpty` check wouldn't be needed, because you made the `selectedVehicle` optional

rramirez-dev commented:
So I should remove "vehicle as Vehicle" from the tag and replace it with "vehicle.id".

mikaelacaron commented:
> So I should remove `vehicle as Vehicle` from the tag and replace it with `vehicle.id`

Yes because it's not necessary to cast the vehicle as it's own type, the goal is to use the `id` as the main point of data for the picker, like you had it before

rramirez-dev committed:
PR-182-Review

-Per PR review I have refactor the vehicle picker to use the vehicle id instead of the vehicle object.
-Replaced "vehicle as Vehicle) for the text tag and replaced it with vehicle.id
-Removed !vehicle.isEmpty check

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift:
```diff
@@ -34,9 +34,9 @@ struct AddMaintenanceView: View {
-                        ForEach(vehicles) { vehicle in
-                            Text(vehicle.name).tag(vehicle as Vehicle)
+                    Picker(selection: $selectedVehicleID) {
+                        ForEach(vehicles, id: \.self) { vehicle in
```
```suggestion
                        ForEach(vehicles) { vehicle in
```

This is not needed, it's about if the elements in the `ForEach` are `Identifiable`, which they are, the `id` doesn't need to be explicitly noted

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift:
```diff
@@ -67,23 +67,21 @@ struct AddMaintenanceView: View {
-                            addTapped(event)
+                        if let selectedVehicleID {
+                            if let vehicle = vehicles.filter({ $0.id == selectedVehicleID }).first {
+                                let event = MaintenanceEvent(title: 
+                                                            title,
+                                                            date: date,
+                                                            notes: notes,
+                                                            vehicle: vehicle)
```
Going to push a change to fix the alignment of this

mikaelacaron reviewed (approved):
Good work! Just a small comment and I can merge this

mikaelacaron committed:
Update Basic-Car-Maintenance/Shared/Dashboard/Views/AddMaintenanceView.swift

mikaelacaron committed:
Fix alignment

mikaelacaron committed:
Merge branch 'dev' into 182-Add-MaintainanceView-Fix
